### PR TITLE
Feature/google analytics

### PR DESCRIPTION
**Pull request by davidebukali**
Fix for issue #42

### CODE REVIEW

File: components/MapSection/MapSection.tsx [hunk 1]
```diff
@@ -69,10 +70,18 @@ const MapSection: FunctionComponent<MapSectionProps> = ({ onChangeLocation, ...p
...
+      topic: optns.theme?.name,
+      indicator: optns.indicator?.name,
+      year: optns.year
+    });
   };
   const onSelectLocation = (location?: SpotlightLocation): void => {
     setQuery(options, location && [location]);
     setActiveLocation(location);
+    addEvent('locationChanged', {
+      locationName: location?.name.toUpperCase()
```
Comment (edwinmp) [inline components/MapSection/MapSection.tsx:83]: Shouldn't this at least default to the country name for when there's no location?

Comment (edwinmp) [reply components/MapSection/MapSection.tsx:83]: Write it on one line ... sure it'll fit.

File: components/MapSection/MapSection.tsx [hunk 2]
```diff
@@ -69,10 +70,18 @@ const MapSection: FunctionComponent<MapSectionProps> = ({ onChangeLocation, ...p
   const onOptionsChange = (optns: SpotlightOptions): void => {
     setQuery(optns, activeLocation && [activeLocation]);
     setOptions(optns);
+    addEvent('mapsectionOptionsChanged', {
```
Comment (edwinmp) [inline components/MapSection/MapSection.tsx:73]: proper camel case ... u know I'm a sucker for such things... consistency.

File: components/LocationComparisonSection/LocationComparisonSection.tsx [hunk 3]
```diff
@@ -55,12 +64,13 @@ const LocationComparisonSection: FunctionComponent<P> = ({ themes, ...props }) =
...
         <LocationComparisonBanner onCompare={onCompare} locations={locations} />
       </PageSection>
-      {chartIDs.map(key => (
+      {chartIDs.map((key, index) => (
         <LocationComparisonChartSection
           key={key}
           themes={themes}
           locations={locations}
           onRemove={chartIDs.length > 1 ? onRemove(key) : undefined}
+          chartNumber={index + 1}
```
Comment (edwinmp) [inline components/LocationComparisonSection/LocationComparisonSection.tsx:73]: We won't need this.

File: components/LocationComparisonSection/LocationComparisonSection.tsx [hunk 4]
```diff
@@ -42,6 +43,14 @@ const LocationComparisonSection: FunctionComponent<P> = ({ themes, ...props }) =
...
       setChartIDs(chartIDs.concat(generateUniqueRandomID([])));
     }
+    const locationString = locations
+      .map(item => {
+        return item.name;
+      })
+      .join(', ');
+    addEvent('locationsCompared', {
+      locations: locationString
+    });
```
Comment (edwinmp) [inline components/LocationComparisonSection/LocationComparisonSection.tsx:53]: Rewrite in one line

File: components/LocationComparisonSection/LocationComparisonChartSection.tsx [hunk 5]
```diff
@@ -28,6 +30,11 @@ const LocationComparisonChartSection: FunctionComponent<ComponentProps> = props
     if (options.indicator) {
       setSelections(options);
       setQuery(options, props.locations);
+      addEvent('locationComparisonOptionsChanged', {
+        topic: options.theme?.name,
+        indicator: options.indicator.name,
+        chartNumber: props.chartNumber
```
Comment (edwinmp) [inline components/LocationComparisonSection/LocationComparisonChartSection.tsx:36]: We don't need this. We only need to check on the topic & indicator... 
image
As you can see, tracking chart number splits the analytics based on chart count, which is a stat we don't care about ... We just want to know which topic & indicator.

File: components/LocationComparisonSection/LocationComparisonChartSection.tsx [hunk 6]
```diff
@@ -10,11 +10,13 @@ import { Spotlight } from '../Spotlight';
 import { SpotlightBanner } from '../SpotlightBanner';
 import { SpotlightFilters } from '../SpotlightFilters';
 import { VisualisationSectionMain } from '../VisualisationSection';
+import { addEvent } from '../../utils/gtm';
 
 interface ComponentProps {
   themes: SpotlightTheme[];
   locations: SpotlightLocation[];
   onRemove?: () => void;
+  chartNumber: number;
```
Comment (edwinmp) [inline components/LocationComparisonSection/LocationComparisonChartSection.tsx:19]: We don't need this.

Comment (davidebukali) [reply components/LocationComparisonSection/LocationComparisonChartSection.tsx:19]: Removed this prop

File: components/LocationComparisonSection/LocationComparisonChartSection.tsx (hunk 5 above)
Comment (davidebukali) [reply components/LocationComparisonSection/LocationComparisonChartSection.tsx:36]: Updated to ignore the chart count

File: components/LocationComparisonSection/LocationComparisonSection.tsx (hunk 4 above)
Comment (davidebukali) [reply components/LocationComparisonSection/LocationComparisonSection.tsx:53]: Updated

File: components/LocationComparisonSection/LocationComparisonSection.tsx (hunk 3 above)
Comment (davidebukali) [reply components/LocationComparisonSection/LocationComparisonSection.tsx:73]: Removed this prop

File: components/MapSection/MapSection.tsx (hunk 2 above)
Comment (davidebukali) [reply components/MapSection/MapSection.tsx:73]: Updated to camel case

File: components/MapSection/MapSection.tsx (hunk 1 above)
Comment (davidebukali) [reply components/MapSection/MapSection.tsx:83]: Updated to default to country name when there's no location which makes it too long for one line.

File: components/RevenueExpenditureSection/RevenueExpenditureSection.tsx [hunk 7]
```diff
@@ -42,6 +43,27 @@ const renderPaddedAlert = (message: string): ReactNode => (
...
+  year: number | string | undefined
+): void => {
+  const eventName =
+    sectionHeading.indexOf('Revenue') !== -1
+      ? 'revenueAndGrantsChartOptionChanged'
+      : sectionHeading.indexOf('Expenditure') !== -1
+      ? 'expenditureChartOptionChanged'
+      : sectionHeading.indexOf('Financing') !== -1
+      ? 'financingChartOptionChanged'
+      : '';
```
[the remark this answers is not in the document]
Comment (davidebukali) [reply components/RevenueExpenditureSection/RevenueExpenditureSection.tsx:59]: Updated to `if else` statements

File: components/RevenueExpenditureSection/RevenueExpenditureSection.tsx [hunk 8]
```diff
@@ -66,6 +88,9 @@ const RevenueExpenditureSection: FunctionComponent<RevenueSectionProps> = ({ ind
     }
   };
 
+  const sectionHeading = processTemplateString(indicator.name, {
+    location: location ? toCamelCase(location.name) : countryName
+  });
```
[the remark this answers is not in the document]
Comment (davidebukali) [reply components/RevenueExpenditureSection/RevenueExpenditureSection.tsx:93]: Its not used in any, moved after useEffect

File: pages/spotlight/[slug]/index.tsx [hunk 9]
```diff
@@ -30,6 +31,11 @@ const Spotlight: NextPage<SpotlightProps> = ({ setData, scaffold, page }) => {
   const [countryInfo] = useState({ countryCode, countryName, currencyCode });
   const [location, setLocation] = useState<SpotlightLocation | undefined>();
   useEffect(() => {
+    window.dataLayer = window.dataLayer || [];
+    addEvent('registrationComplete', {
+      registrationCountry: 'United States',
+      plan: 'Premiums yo'
+    });
```
[the remark this answers is not in the document]
Comment (davidebukali) [reply pages/spotlight/[slug]/index.tsx:38]: Thats the first example code; Just removed it

File: pages/spotlight/[slug]/index.tsx [hunk 10]
```diff
@@ -30,6 +31,11 @@ const Spotlight: NextPage<SpotlightProps> = ({ setData, scaffold, page }) => {
   const [countryInfo] = useState({ countryCode, countryName, currencyCode });
   const [location, setLocation] = useState<SpotlightLocation | undefined>();
   useEffect(() => {
+    window.dataLayer = window.dataLayer || [];
```
[the remark this answers is not in the document]
Comment (davidebukali) [reply pages/spotlight/[slug]/index.tsx:34]: Indeed it belongs on the abstracted file; I have moved it to `addEvent`

File: utils/gtm.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
+export {};
+
+declare global {
+  interface Window {
+    dataLayer: any;
+  }
+}
+
+export const addEvent = (eventName: string, data: any): void => {
```
[the remark this answers is not in the document]
Comment (davidebukali) [reply utils/gtm.ts:9]: Updated to an empty object

File: utils/gtm.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+export {};
+
+declare global {
+  interface Window {
+    dataLayer: any;
+  }
+}
```
[the remark this answers is not in the document]
Comment (davidebukali) [reply utils/gtm.ts:7]: Updated

File: utils/gtm.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
+export {};
```
[the remark this answers is not in the document]
Comment (davidebukali) [reply utils/gtm.ts:1]: Removed unnecessary export

File: components/IndicatorComparisonSection/IndicatorSelectionBanner.tsx [hunk 14]
```diff
@@ -3,6 +3,7 @@ import { SpotlightOptions, SpotlightTheme, getDefaultsByIndex } from '../../util
 import { SpotlightBanner } from '../SpotlightBanner';
 import { SpotlightComparison } from '../SpotlightComparison';
 import { SpotlightFilters } from '../SpotlightFilters';
+import { addEvent } from '../../utils/gtm';
```
[the remark this answers is not in the document]
Comment (davidebukali) [reply components/IndicatorComparisonSection/IndicatorSelectionBanner.tsx:6]: Renamed to analytics and updated google tag manager with appropriate categories, actions and labels.

File: utils/context.ts [hunk 15]
```diff
@@ -1,6 +1,7 @@
-import { createContext } from 'react';
+import { createContext, useContext } from 'react';
 import { SpotlightLocation } from '.';
 import { CountryInfo } from './data';
 
 export const LocationContext = createContext<SpotlightLocation | undefined>(undefined);
 export const CountryContext = createContext<CountryInfo>({ countryCode: '', countryName: '', currencyCode: '' });
+export const useCountryContext = (): CountryInfo => useContext(CountryContext);
```
Comment (edwinmp) [inline utils/context.ts:7]: Nope ... not a very useful change ... wherever u want to use the CountryContext, write this line instead ...

File: components/RevenueExpenditureSection/RevenueExpenditureSection.tsx [hunk 16]
```diff
@@ -79,19 +103,33 @@ const RevenueExpenditureSection: FunctionComponent<RevenueSectionProps> = ({ ind
       setYearBudgetTypes();
     }
   }, [dataLoading]);
+  const countryContext = useCountryContext();
```
Comment (edwinmp) [inline components/RevenueExpenditureSection/RevenueExpenditureSection.tsx:106]: There's already a country context on line 71, nicely destructured ...

File: components/MapSection/MapSection.tsx [hunk 17]
```diff
@@ -65,20 +66,35 @@ const MapSection: FunctionComponent<MapSectionProps> = ({ onChangeLocation, ...p
       onChangeLocation(activeLocation);
     }
   }, []);
+  const countryContext = useCountryContext();
```
Comment (edwinmp) [inline components/MapSection/MapSection.tsx:69]: Use the actual hook here (useContext), and move it before the useEffect.

File: components/LocationComparisonSection/LocationComparisonSection.tsx [hunk 18]
```diff
@@ -42,6 +43,8 @@ const LocationComparisonSection: FunctionComponent<P> = ({ themes, ...props }) =
     if (!chartIDs || chartIDs.length === 0) {
       setChartIDs(chartIDs.concat(generateUniqueRandomID([])));
     }
+    const locationString = locations.map(item => item.name).join(', ');
+    addEvent('locationsCompared', { locations: locationString });
```
Comment (edwinmp) [inline components/LocationComparisonSection/LocationComparisonSection.tsx:47]: You're not reusing locationString anywhere else, so may as well inline it.

File: components/IndicatorComparisonSection/IndicatorSelectionBanner.tsx [hunk 19]
```diff
@@ -20,6 +21,7 @@ const IndicatorSelectionBanner: FunctionComponent<ComparisonWrapperProps> = ({ t
       onCompare([getDefaultsByIndex(themes).selected, getDefaultsByIndex(themes, DEFAULT_INDEXES).selected]);
     }
   }, []);
+  const countryContext = useCountryContext();
```
Comment (edwinmp) [inline components/IndicatorComparisonSection/IndicatorSelectionBanner.tsx:24]: Use the actual hook here ('useContext') and move it before the useEffect.

File: utils/context.ts (hunk 15 above)
Comment (edwinmp) [reply utils/context.ts:7]: Or maybe not ... hmm... oh well, just leave it ... but try to check if there's already an existing usage for CountryContext. Don't mind the other comments about it besides moving them.

File: components/LocationComparisonSection/LocationComparisonSection.tsx (hunk 18 above)
Comment (davidebukali) [reply components/LocationComparisonSection/LocationComparisonSection.tsx:47]: Now inlined

File: components/RevenueExpenditureSection/RevenueExpenditureSection.tsx (hunk 16 above)
Comment (davidebukali) [reply components/RevenueExpenditureSection/RevenueExpenditureSection.tsx:106]: Indeed, now using destructured countryName

**Pull request merged**